File: frontend/src/shared/types/media.ts
import { MediaGenre } from './genre';

export abstract class Media {
  id: string = '';
  title: string | null;
  description: string | null;
  genreList: MediaGenre[] | null;

  isComplete: boolean;

  rating: number | null;
  watched: boolean;

  durationWatched: number;
  duration: number;

  abstract type: string;

  constructor(data: Partial<Media>) {
    this.id = data.id ?? '';
    this.title = data.title ?? null;
    this.description = data.description ?? null;
    this.genreList = data.genreList ?? null;
    this.rating = data.rating ?? null;
    this.isComplete = data.isComplete ?? true;
    this.watched = data.watched ?? false;
    this.duration = data.duration ?? 0;
    this.durationWatched = data.durationWatched ?? 0;
  }

  public abstract getMediaType(): string;
}

File: backend/classes/media.py
from typing import Optional
from pydantic import BaseModel, Field
from classes.media_genre import MediaGenre


class Media(BaseModel):
    id: str

    title: str | None = None
    description: str | None = None

    genreList: list[MediaGenre] = []

    type: str = ""

    isComplete: bool = True

    durationWatched: int = 0
    duration: int = 0

    rating: int | None = None

    watched: bool = False

    mediaPath: str
